File: c/3-pid-bm/main/controller.c
#include <math.h>

#include "driver/gptimer.h"
#include "driver/ledc.h"
#include "esp_adc/adc_oneshot.h"
#include "esp_log.h"
#include "freertos/idf_additions.h"
#include "freertos/task.h"
#include "hal/ledc_types.h"
#include "mb_endianness_utils.h"
#include "portmacro.h"

#include "controller.h"
#include "perf.h"
#include "ringbuffer.h"

// Configuration
static const adc_unit_t ADC_UNIT = ADC_UNIT_1;
static const adc_channel_t ADC_CHANNEL = ADC_CHANNEL_4;
static const adc_atten_t ADC_ATTENUATION = ADC_ATTEN_DB_12;
static const adc_bitwidth_t ADC_BITWIDTH = ADC_BITWIDTH_9;

static const uint32_t PWM_SPEED = LEDC_LOW_SPEED_MODE;
static const uint32_t PWM_CHANNEL = LEDC_CHANNEL_0;
static const uint32_t PWM_GPIO = 5;
static const uint32_t PWM_FREQUENCY = 1000;
static const uint32_t PWM_DUTY_RESOLUTION = LEDC_TIMER_13_BIT;
static const float PWM_MIN = 0.10;
static const float PWM_MAX = 1.00;
static const float PWM_LIMIT_MIN_DEADZONE = 0.001;

static const uint32_t TIMER_FREQUENCY = 1000000; // period = 1us
static const size_t CONTROL_ITERS_PER_PERF_REPORT = 10;

// Derived constants
static const uint32_t ADC_MAX_VALUE = (1 << ADC_BITWIDTH) - 1;
static const uint32_t PWM_DUTY_MAX = (1 << PWM_DUTY_RESOLUTION) - 1;

static const char TAG[] = "controller";

TaskHandle_t controller_task = NULL;

esp_err_t read_adc(controller_t *self, float *value) {
  int value_raw;

  esp_err_t err = adc_oneshot_read(self->adc, ADC_CHANNEL, &value_raw);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "adc_oneshot_read fail (0x%x)", err);
    return err;
  }

  *value = (float)value_raw / ADC_MAX_VALUE;
  return ESP_OK;
}

esp_err_t set_duty_cycle(float value) {
  esp_err_t err;

  const uint32_t duty_cycle = value * PWM_DUTY_MAX;

  err = ledc_set_duty(PWM_SPEED, PWM_CHANNEL, duty_cycle);
  ESP_ERROR_CHECK_WITHOUT_ABORT(err);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "ledc_set_duty fail (0x%x)", err);
    return err;
  }

  err = ledc_update_duty(PWM_SPEED, PWM_CHANNEL);
  ESP_ERROR_CHECK_WITHOUT_ABORT(err);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "ledc_update_duty fail (0x%x)", err);
    return err;
  }

  return ESP_OK;
}

bool IRAM_ATTR on_timer_fired(
    gptimer_handle_t timer, const gptimer_alarm_event_data_t *edata,
    void *user_data
) {
  BaseType_t high_task_awoken = pdFALSE;
  vTaskNotifyGiveFromISR(controller_task, &high_task_awoken);
  return high_task_awoken == pdTRUE;
}

esp_err_t controller_init(
    controller_t *self, registers_t *registers, controller_options_t options
) {
  esp_err_t err;

  ringbuffer_t *revolutions;
  err = ringbuffer_init(&revolutions, options.revolution_bins);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "ringbuffer_init fail (0x%x)", err);
    return err;
  }

  adc_oneshot_unit_handle_t adc;
  adc_oneshot_unit_init_cfg_t init_config1 = {.unit_id = ADC_UNIT};
  err = adc_oneshot_new_unit(&init_config1, &adc);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "adc_oneshot_new_unit fail (0x%x)", err);
    ringbuffer_deinit(revolutions);
    return err;
  }

  adc_oneshot_chan_cfg_t config = {
      .atten = ADC_ATTENUATION,
      .bitwidth = ADC_BITWIDTH,
  };
  err = adc_oneshot_config_channel(adc, ADC_CHANNEL, &config);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "adc_oneshot_config_channel fail (0x%x)", err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(adc_oneshot_del_unit(adc));
    ringbuffer_deinit(revolutions);
    return err;
  }

  const ledc_timer_config_t led_timer_config = {
      .speed_mode = PWM_SPEED,
      .duty_resolution = PWM_DUTY_RESOLUTION,
      .timer_num = LEDC_TIMER_0,
      .freq_hz = PWM_FREQUENCY,
      .clk_cfg = LEDC_AUTO_CLK,
  };
  err = ledc_timer_config(&led_timer_config);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "ledc_timer_config fail (0x%x)", err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(adc_oneshot_del_unit(adc));
    ringbuffer_deinit(revolutions);
    return err;
  }

  ledc_channel_config_t ledc_channel = {
      .speed_mode = PWM_SPEED,
      .channel = PWM_CHANNEL,
      .timer_sel = led_timer_config.timer_num,
      .intr_type = LEDC_INTR_DISABLE,
      .gpio_num = PWM_GPIO,
      .duty = 0,
      .hpoint = 0
  };
  err = ledc_channel_config(&ledc_channel);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "ledc_channel_config fail (0x%x)", err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(adc_oneshot_del_unit(adc));
    ringbuffer_deinit(revolutions);
    return err;
  }

  gptimer_handle_t timer;
  err = gptimer_new_timer(
      &(gptimer_config_t){
          .clk_src = GPTIMER_CLK_SRC_DEFAULT,
          .direction = GPTIMER_COUNT_UP,
          .resolution_hz = TIMER_FREQUENCY,
      },
      &timer
  );
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "timer_init fail (0x%x)", err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(adc_oneshot_del_unit(adc));
    ringbuffer_deinit(revolutions);
    return err;
  }

  err = gptimer_register_event_callbacks(
      timer,
      &(gptimer_event_callbacks_t){
          .on_alarm = on_timer_fired,
      },
      NULL
  );
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "gptimer_register_event_callbacks fail (0x%x)", err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(gptimer_del_timer(timer));
    ESP_ERROR_CHECK_WITHOUT_ABORT(adc_oneshot_del_unit(adc));
    ringbuffer_deinit(revolutions);
    return err;
  }
  err = gptimer_enable(timer);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "gptimer_enable fail (0x%x)", err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(gptimer_del_timer(timer));
    ESP_ERROR_CHECK_WITHOUT_ABORT(adc_oneshot_del_unit(adc));
    ringbuffer_deinit(revolutions);
    return err;
  }
  err = gptimer_set_alarm_action(
      timer,
      &(gptimer_alarm_config_t){
          .alarm_count = TIMER_FREQUENCY / options.frequency,
          .reload_count = 0,
          .flags.auto_reload_on_alarm = true,
      }
  );
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "gptimer_set_alarm_action fail (0x%x)", err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(gptimer_disable(timer));
    ESP_ERROR_CHECK_WITHOUT_ABORT(gptimer_del_timer(timer));
    ESP_ERROR_CHECK_WITHOUT_ABORT(adc_oneshot_del_unit(adc));
    ringbuffer_deinit(revolutions);
    return err;
  }

  const float interval_rotate_once_s =
      (float)1 / options.frequency * options.reads_per_bin;
  const float interval_rotate_all_s =
      interval_rotate_once_s * options.revolution_bins;

  *self = (controller_t){
      .options = options,
      .registers = registers,
      .adc = adc,
      .timer = timer,
      .interval =
          {
              .rotate_once_s = interval_rotate_once_s,
              .rotate_all_s = interval_rotate_all_s,
          },
      .state =
          {
              .revolutions = revolutions,
              .is_close = false,
              .feedback = {.delta = 0, .integration_component = 0},
          },
  };

  return ESP_OK;
}

void controller_deinit(controller_t *self) {
  free(self->state.revolutions);
  ESP_ERROR_CHECK_WITHOUT_ABORT(gptimer_stop(self->timer));
  ESP_ERROR_CHECK_WITHOUT_ABORT(gptimer_disable(self->timer));
  ESP_ERROR_CHECK_WITHOUT_ABORT(gptimer_del_timer(self->timer));
  ESP_ERROR_CHECK_WITHOUT_ABORT(ledc_stop(PWM_SPEED, PWM_CHANNEL, 0));
  ESP_ERROR_CHECK_WITHOUT_ABORT(adc_oneshot_del_unit(self->adc));
  ringbuffer_deinit(self->state.revolutions);
}

float finite_or_zero(float value) { return isfinite(value) ? value : 0; }

float limit(float value, float min, float max) {
  if (value < PWM_LIMIT_MIN_DEADZONE)
    return 0;

  const float result = value + min;
  return result < min ? min : (result > max ? max : result);
}

float calculate_frequency(controller_t *self) {
  uint32_t sum = 0;
  for (size_t i = 0; i < self->state.revolutions->length; ++i)
    sum += self->state.revolutions->array[i];

  return (float)sum / self->interval.rotate_all_s;
}

typedef struct {
  float target_frequency;
  float proportional_factor;
  float integration_time;
  float differentiation_time;
} control_params_t;

control_params_t read_control_params(controller_t *self) {
  registers_holding_t *holding = &self->registers->holding;
  return (control_params_t){
      .target_frequency = mb_get_float_cdab(&holding->target_frequency),
      .proportional_factor = mb_get_float_cdab(&holding->proportional_factor),
      .integration_time = mb_get_float_cdab(&holding->integration_time),
      .differentiation_time = mb_get_float_cdab(&holding->differentiation_time),
  };
}

typedef struct {
  float signal;
  feedback_t feedback;
} control_t;

control_t calculate_control(
    controller_t *self, const control_params_t *params, float frequency
) {
  const float interval_s = self->interval.rotate_once_s;

  const float integration_factor =
      params->proportional_factor / params->integration_time * interval_s;
  const float differentiation_factor =
      params->proportional_factor * params->differentiation_time / interval_s;

  const float delta = params->target_frequency - frequency;
  ESP_LOGD(TAG, "delta: %.2f", delta);

  const float proportional_component = params->proportional_factor * delta;
  const float integration_component =
      self->state.feedback.integration_component +
      integration_factor * self->state.feedback.delta;
  const float differentiation_component =
      differentiation_factor * (delta - self->state.feedback.delta);

  const float control_signal = proportional_component + integration_component +
                               differentiation_component;
  ESP_LOGD(
      TAG, "control_signal: %.2f = %.2f + %.2f + %.2f", control_signal,
      proportional_component, integration_component, differentiation_component
  );

  return (control_t
  ){.signal = control_signal,
    .feedback = {
        .delta = finite_or_zero(delta),
        .integration_component = finite_or_zero(integration_component),
    }};
}

void write_state(controller_t *self, float frequency, float control_signal) {
  registers_input_t *input = &self->registers->input;
  mb_set_float_cdab(&input->frequency, frequency);
  mb_set_float_cdab(&input->control_signal, control_signal);
}

esp_err_t read_phase(controller_t *self) {
  esp_err_t err;

  float value;
  err = read_adc(self, &value);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "read_adc fail (0x%x)", err);
    return err;
  }

  if (value < self->options.revolution_treshold_close &&
      !self->state.is_close) {
    // gone close
    self->state.is_close = true;
    *ringbuffer_back(self->state.revolutions) += 1;
  } else if (value > self->options.revolution_treshold_far &&
             self->state.is_close) {
    // gone far
    self->state.is_close = false;
  }

  return ESP_OK;
}

esp_err_t control_phase(controller_t *self) {
  esp_err_t err;

  const float frequency = calculate_frequency(self);
  ringbuffer_push(self->state.revolutions, 0);

  const control_params_t params = read_control_params(self);

  const control_t control = calculate_control(self, &params, frequency);

  const float control_signal_limited = limit(control.signal, PWM_MIN, PWM_MAX);
  ESP_LOGD(TAG, "control_signal_limited: %.2f", control_signal_limited);

  write_state(self, frequency, control_signal_limited);
  err = set_duty_cycle(control_signal_limited);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "set_duty_cycle fail (0x%x)", err);
    return err;
  }

  self->state.feedback = control.feedback;

  ESP_LOGD(TAG, "frequency: %.2f", frequency);

  return ESP_OK;
}

void controller_loop(void *params) {
  esp_err_t err;
  controller_t *self = params;

  ESP_LOGI(TAG, "Starting controller");

  if (controller_task != NULL) {
    ESP_LOGE(TAG, "controller task already running");
    abort();
  }
  controller_task = xTaskGetCurrentTaskHandle();

  err = gptimer_start(self->timer);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "gptimer_start fail (0x%x)", err);
    abort();
  }

  perf_counter_t perf_read;
  err = perf_counter_init(&perf_read, "READ");
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "perf_counter_init (READ) fail (0x%x)", err);
    abort();
  }

  perf_counter_t perf_control;
  err = perf_counter_init(&perf_control, "CONTROL");
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "perf_counter_init (CONTROL) fail (0x%x)", err);
    abort();
  }

  while (true) {
    for (size_t i = 0; i < CONTROL_ITERS_PER_PERF_REPORT; ++i) {
      for (size_t j = 0; j < self->options.reads_per_bin; ++j) {
        while (ulTaskNotifyTake(pdTRUE, portMAX_DELAY) == 0)
          ;

        const perf_start_mark_t read_start = perf_counter_mark_start();

        err = read_phase(self);
        if (err != ESP_OK)
          ESP_LOGE(TAG, "read_phase fail (0x%x)", err);

        perf_counter_add_sample(&perf_read, read_start);
      }

      const perf_start_mark_t control_start = perf_counter_mark_start();

      err = control_phase(self);
      if (err != ESP_OK)
        ESP_LOGE(TAG, "control_phase fail (0x%x)", err);

      perf_counter_add_sample(&perf_control, control_start);
    }
    perf_counter_report(&perf_read);
    perf_counter_report(&perf_control);

    perf_counter_reset(&perf_read);
    perf_counter_reset(&perf_control);
  }

  controller_task = NULL;
}

File: zig/3-pid-bm/main/Controller.zig
const std = @import("std");
const idf = @import("esp_idf");
const sys = idf.sys;
const Allocator = std.mem.Allocator;

const Registers = @import("Registers.zig");
const RingBuffer = @import("ring_buffer.zig").RingBuffer;
const perf_m = @import("perf.zig");
const memory = @import("memory.zig");
const c = @import("c.zig");
const utils = @import("utils.zig");
const logErr = utils.logErr;
const panicErr = utils.panicErr;

const adc_unit = c.ADC_UNIT_1;
const adc_channel = c.ADC_CHANNEL_4;
const adc_bitwidth = c.ADC_BITWIDTH_9;
const adc_max = (1 << adc_bitwidth) - 1;
const adc_channel_config = c.adc_oneshot_chan_cfg_t{
    .atten = c.ADC_ATTEN_DB_12,
    .bitwidth = adc_bitwidth,
};

const pwm_timer = c.LEDC_TIMER_0;
const pwm_speed_mode = c.LEDC_LOW_SPEED_MODE;
const pwm_channel = c.LEDC_CHANNEL_0;
const pwm_bitwidth = c.LEDC_TIMER_13_BIT;
const pwm_duty_max = (1 << pwm_bitwidth) - 1;
const pwm_timer_config = c.ledc_timer_config_t{
    .timer_num = pwm_timer,
    .speed_mode = pwm_speed_mode,
    .duty_resolution = pwm_bitwidth,
    .freq_hz = 1000,
    .clk_cfg = c.LEDC_AUTO_CLK,
};
const pwm_timer_deconfig = c.ledc_timer_config_t{
    .timer_num = pwm_timer,
    .deconfigure = true,
};
const pwm_channel_config = c.fixed.ledc_channel_config_t{
    .timer_sel = pwm_timer,
    .channel = pwm_channel,
    .speed_mode = pwm_speed_mode,
    .intr_type = c.LEDC_INTR_DISABLE,
    .gpio_num = c.GPIO_NUM_5,
    .duty = 0,
    .hpoint = 0,
};

const pwm_min = 0.1;
const pwm_max = 1.0;
const limit_min_deadzone = 0.001;

const timer_frequency = 1000000; // period = 1us
const control_iters_per_perf_report = 10;

var controller_task: sys.TaskHandle_t = null;

allocator: Allocator,
options: InitOptions,
registers: *Registers,
adc: c.adc_oneshot_unit_handle_t,
timer: c.gptimer_handle_t,
interval: struct {
    rotate_once_s: f32,
    rotate_all_s: f32,
},
state: struct {
    revolutions: RingBuffer(u32),
    is_close: bool,
    feedback: Feedback,
},

pub const InitOptions = struct {
    /// Frequency of control phase, during which the following happens:
    /// * calculating the frequency for the current time window,
    /// * moving the time window forward,
    /// * updating duty cycle,
    /// * updating modbus registers.
    control_frequency: u32,
    /// Frequency is estimated for the current time window. That window is broken
    /// into [time_window_bins] bins and is moved every time the control phase
    /// takes place.
    time_window_bins: u32,
    /// Each bin in the time window gets [reads_per_bin] reads, before the next
    /// control phase fires. That means, that the read phase occurs with frequency
    /// equal to:
    ///     `control_frequency * reads_per_bin`
    /// , because every time the window moves (control phase), there must be
    /// [reads_per_bin] reads in the last bin already (read phase).
    reads_per_bin: u32,
    /// When ADC reads below this signal, the state is set to `close` to the
    /// motor magnet. If the state has changed, a new revolution is counted.
    revolution_treshold_close: f32,
    /// When ADC reads above this signal, the state is set to `far` from the
    /// motor magnet.
    revolution_treshold_far: f32,
};

const Feedback = struct {
    delta: f32 = 0,
    integration_component: f32 = 0,
};

const Self = @This();

pub fn init(allocator: Allocator, registers: *Registers, options: InitOptions) !Self {
    const revolutions = try RingBuffer(u32).init(allocator, options.time_window_bins);
    errdefer revolutions.deinit(allocator);

    var adc: c.adc_oneshot_unit_handle_t = undefined;
    try c.espCheckError(c.adc_oneshot_new_unit(&.{ .unit_id = adc_unit }, &adc));
    errdefer c.espLogError(c.adc_oneshot_del_unit(adc), "adc_oneshot_del_unit");

    try c.espCheckError(c.adc_oneshot_config_channel(adc, adc_channel, &adc_channel_config));

    try c.espCheckError(c.ledc_timer_config(&pwm_timer_config));
    errdefer c.espLogError(c.ledc_timer_config(&pwm_timer_deconfig), "ledc_timer_config");

    try c.espCheckError(c.fixed.ledc_channel_config(&pwm_channel_config));
    errdefer c.espLogError(c.ledc_stop(pwm_speed_mode, pwm_channel, 0), "ledc_stop");

    var timer: c.gptimer_handle_t = undefined;
    try c.espCheckError(gptimer_new_timer(
        &.{
            .clk_src = c.GPTIMER_CLK_SRC_DEFAULT,
            .direction = c.GPTIMER_COUNT_UP,
            .resolution_hz = timer_frequency,
        },
        &timer,
    ));
    errdefer c.espLogError(c.gptimer_del_timer(timer), "gptimer_del_timer");

    try c.espCheckError(c.gptimer_register_event_callbacks(
        timer,
        &.{ .on_alarm = onTimerFired },
        null,
    ));

    try c.espCheckError(c.gptimer_enable(timer));
    errdefer c.espLogError(c.gptimer_disable(timer), "gptimer_disable");

    const read_frequency = options.control_frequency * options.reads_per_bin;
    try c.espCheckError(gptimer_set_alarm_action(timer, &.{
        .alarm_count = timer_frequency / read_frequency,
        .reload_count = 0,
        .flags = .{ .auto_reload_on_alarm = true },
    }));

    const interval_rotate_once_s: f32 =
        1.0 / @as(f32, @floatFromInt(options.control_frequency));
    const interval_rotate_all_s: f32 =
        interval_rotate_once_s * @as(f32, @floatFromInt(options.time_window_bins));

    return .{
        .allocator = allocator,
        .options = options,
        .registers = registers,
        .adc = adc,
        .timer = timer,
        .interval = .{
            .rotate_once_s = interval_rotate_once_s,
            .rotate_all_s = interval_rotate_all_s,
        },
        .state = .{
            .revolutions = revolutions,
            .is_close = false,
            .feedback = .{},
        },
    };
}

pub fn deinit(self: *const Self) void {
    c.espLogError(c.gptimer_stop(self.timer), "gptimer_stop");
    c.espLogError(c.gptimer_disable(self.timer), "gptimer_disable");
    c.espLogError(c.gptimer_del_timer(self.timer), "gptimer_del_timer");
    c.espLogError(c.ledc_stop(pwm_speed_mode, pwm_channel, 0), "ledc_stop");
    c.espLogError(c.ledc_timer_config(&pwm_timer_deconfig), "ledc_timer_config");
    c.espLogError(c.adc_oneshot_del_unit(self.adc), "adc_oneshot_del_unit");
    self.state.revolutions.deinit(self.allocator);
}

pub fn run(args: ?*anyopaque) callconv(.c) void {
    const self: *Self = @alignCast(@ptrCast(args));

    std.log.info("Starting controller", .{});

    if (controller_task != null)
        std.debug.panic("Controller task already running", .{});
    controller_task = idf.xTaskGetCurrentTaskHandle();
    defer controller_task = null;

    c.espCheckError(c.gptimer_start(self.timer)) catch |err| panicErr(err);

    const control_frequency = self.options.control_frequency;
    const read_frequency = control_frequency * self.options.reads_per_bin;

    var perf_read = perf_m.Counter.init(
        self.allocator,
        "READ",
        read_frequency,
    ) catch |err| panicErr(err);
    defer perf_read.deinit();

    var perf_control = perf_m.Counter.init(
        self.allocator,
        "CONTROL",
        control_frequency,
    ) catch |err| panicErr(err);
    defer perf_control.deinit();

    var report_number: u64 = 0;
    while (true) {
        for (0..self.options.control_frequency) |_| {
            for (0..self.options.reads_per_bin) |_| {
                while (idf.ulTaskGenericNotifyTake(
                    c.tskDEFAULT_INDEX_TO_NOTIFY,
                    c.pdTRUE,
                    c.portMAX_DELAY,
                ) == 0) {}

                const read_start = perf_m.Marker.now();
                logErr(self.readPhase(), "Controller.readPhase");
                perf_read.addSample(read_start);
            }

            const control_start = perf_m.Marker.now();
            logErr(self.controlPhase(), "Controller.controlPhase");
            perf_control.addSample(control_start);
        }

        std.log.info("# REPORT {}", .{report_number});
        memory.report();
        perf_read.report();
        perf_control.report();
        perf_read.reset();
        perf_control.reset();
        report_number += 1;
    }
}

fn readPhase(self: *Self) !void {
    const value = try self.readAdc();

    if (!self.state.is_close and value < self.options.revolution_treshold_close) {
        // gone close
        self.state.is_close = true;
        self.state.revolutions.back().* += 1;
    } else if (self.state.is_close and value > self.options.revolution_treshold_far) {
        // gone far
        self.state.is_close = false;
    }
}

fn readAdc(self: *Self) !f32 {
    var value: c_int = undefined;
    try c.espCheckError(c.adc_oneshot_read(self.adc, adc_channel, &value));
    return @as(f32, @floatFromInt(value)) / adc_max;
}

fn controlPhase(self: *Self) !void {
    const frequency = self.calculateFrequency();
    try self.state.revolutions.push(0);
    std.log.debug("frequency: {d:.2} Hz", .{frequency});

    const control_params = self.readControlParams();

    const control = self.calculateControl(&control_params, frequency);

    const control_signal_limited = limit(control.signal, pwm_min, pwm_max);
    std.log.debug("control signal limited: {d:.2}", .{control_signal_limited});

    try setDutyCycle(control_signal_limited);

    self.writeRegisters(frequency, control_signal_limited);

    self.state.feedback = control.feedback;
}

fn calculateFrequency(self: *Self) f32 {
    var sum: u32 = 0;
    for (self.state.revolutions.items) |value|
        sum += value;

    return @as(f32, @floatFromInt(sum)) / self.interval.rotate_all_s;
}

pub const ControlParams = struct {
    target_frequency: f32,
    proportional_factor: f32,
    integration_time: f32,
    differentiation_time: f32,
};

fn readControlParams(self: *Self) ControlParams {
    const holding = self.registers.holding;
    return .{
        .target_frequency = holding.target_frequency.toF32(),
        .proportional_factor = holding.proportional_factor.toF32(),
        .integration_time = holding.integration_time.toF32(),
        .differentiation_time = holding.differentiation_time.toF32(),
    };
}

fn calculateControl(
    self: *const Self,
    params: *const ControlParams,
    frequency: f32,
) struct { signal: f32, feedback: Feedback } {
    const interval_s = self.interval.rotate_once_s;

    const integration_factor: f32 =
        params.proportional_factor / params.integration_time * interval_s;
    const differentiation_factor: f32 =
        params.proportional_factor * params.differentiation_time / interval_s;

    const delta: f32 = params.target_frequency - frequency;
    std.log.debug("delta: {d:.2}", .{delta});

    const proportional_component: f32 = params.proportional_factor * delta;
    const integration_component: f32 =
        self.state.feedback.integration_component +
        integration_factor * self.state.feedback.delta;
    const differentiation_component: f32 =
        differentiation_factor * (delta - self.state.feedback.delta);

    const control_signal: f32 = proportional_component +
        integration_component +
        differentiation_component;

    std.log.debug("control_signal: {d:.2} = {d:.2} + {d:.2} + {d:.2}", .{
        control_signal,
        proportional_component,
        integration_component,
        differentiation_component,
    });

    return .{
        .signal = control_signal,
        .feedback = .{
            .delta = delta,
            .integration_component = integration_component,
        },
    };
}

fn setDutyCycle(value: f32) !void {
    const duty_cycle: u32 = @intFromFloat(value * pwm_duty_max);
    try c.espCheckError(c.ledc_set_duty(pwm_speed_mode, pwm_channel, duty_cycle));
    try c.espCheckError(c.ledc_update_duty(pwm_speed_mode, pwm_channel));
}

fn finiteOrZero(value: f32) f32 {
    return if (std.math.isFinite(value)) 0 else value;
}

fn limit(value: f32, min: f32, max: f32) f32 {
    if (value < limit_min_deadzone)
        return 0;

    const result = value + min;
    return if (result < min) min else if (result > max) max else result;
}

fn writeRegisters(self: *Self, frequency: f32, control_signal: f32) void {
    var input = &self.registers.input;
    input.frequency = .fromF32(frequency);
    input.control_signal = .fromF32(control_signal);
}

export fn onTimerFired(
    _: c.gptimer_handle_t,
    _: [*c]const c.gptimer_alarm_event_data_t,
    _: ?*anyopaque,
) linksection(".iram1.0") callconv(.c) bool {
    var high_task_awoken: isize = 0;
    _ = idf.vTaskGenericNotifyGiveFromISR(
        controller_task,
        c.tskDEFAULT_INDEX_TO_NOTIFY,
        &high_task_awoken,
    );
    return high_task_awoken != 0;
}

pub const gptimer_config_t = extern struct {
    clk_src: c.gptimer_clock_source_t = @import("std").mem.zeroes(c.gptimer_clock_source_t),
    direction: c.gptimer_count_direction_t = @import("std").mem.zeroes(c.gptimer_count_direction_t),
    resolution_hz: u32 = @import("std").mem.zeroes(u32),
    intr_priority: c_int = @import("std").mem.zeroes(c_int),
    flags: packed struct(u32) {
        intr_shared: bool = false,
        allow_pd: bool = false,
        backup_before_sleep: bool = false,
        _padding: u29 = 0,
    } = .{},
};
pub extern fn gptimer_new_timer(config: ?*const gptimer_config_t, ret_timer: [*c]c.gptimer_handle_t) c.esp_err_t;

pub const gptimer_alarm_config_t = extern struct {
    alarm_count: u64 = @import("std").mem.zeroes(u64),
    reload_count: u64 = @import("std").mem.zeroes(u64),
    flags: packed struct(u32) {
        auto_reload_on_alarm: bool = false,
        _padding: u31 = 0,
    } = .{},
};
pub extern fn gptimer_set_alarm_action(timer: c.gptimer_handle_t, config: ?*const gptimer_alarm_config_t) c.esp_err_t;
